bundler/lib/bundler/definition.rb

```diff
@@ -270,9 +271,15 @@ def specs_for(groups)
 
     def dependencies_for(groups)
       groups.map!(&:to_sym)
-      current_dependencies.reject do |d|
-        (d.groups & groups).empty?
+      deps = current_dependencies # always returns a new array
+      deps.select! do |d|
+        if RUBY_VERSION >= "3.1"
+          d.groups.intersect?(groups)
+        else
+          !(d.groups & groups).empty?
+        end
```

Comment:
Does `reject!` work?

Reply:
Oh so using `intersect?` is also for performance? I thought the performance was just modifying the array in place, and while intersect is cool, didn't seem worth the extra complexity of having two code branches, etc. But it's fine either way, specially if this way is faster.

Reply:
`intersect?` saves the array allocation from `&`

Reply:
Is it worth calculating which version to use outside the block? Is the RUBY_VERSION check is doing anything complex like splitting strings?